flake.nix

```diff
@@ -5,46 +5,41 @@
 
     let
       forAllSystems = f: nixpkgs.lib.genAttrs systems (system: f system);
-      systems = [ "x86_64-linux" "aarch64-linux" "i686-linux" "x86_64-darwin" ];
+      systems = [
+        "x86_64-linux"
+        "aarch64-linux"
+        "i686-linux"
+        "x86_64-darwin"
+        "aarch64-darwin"
+      ];
     in {
 
       overlay = final: prev: {
-
-        nix-serve = with final; stdenv.mkDerivation {
-          name = "nix-serve-${self.lastModifiedDate}";
-
-          buildInputs = [ perl nixVersions.latest.perl-bindings perlPackages.Plack perlPackages.Starman perlPackages.DBDSQLite ];
-
-          unpackPhase = "true";
-
-          installPhase =
-            ''
-              mkdir -p $out/libexec/nix-serve
-              cp ${./nix-serve.psgi} $out/libexec/nix-serve/nix-serve.psgi
-
-              mkdir -p $out/bin
-              cat > $out/bin/nix-serve <<EOF
-              #! ${stdenv.shell}
-              PERL5LIB=$PERL5LIB \
-              NIX_REMOTE="\''${NIX_REMOTE:-auto?path-info-cache-size=0}" \
-              exec ${perlPackages.Starman}/bin/starman --preload-app $out/libexec/nix-serve/nix-serve.psgi "\$@"
-              EOF
-              chmod +x $out/bin/nix-serve
-            '';
+        nix-serve = final.pkgs.callPackage ./package.nix {
+          inherit self;
+          nix = final.nixVersions.git;
         };
-
       };
 
-      packages = forAllSystems (system: {
-        nix-serve = (import nixpkgs { inherit system; overlays = [ self.overlay ]; }).nix-serve;
+      packages = forAllSystems (system: let
+        pkgs = nixpkgs.legacyPackages.${system};
+      in {
+        nix-serve = nixpkgs.legacyPackages.${system}.callPackage ./package.nix {
+          inherit self;
+          nix = pkgs.nixVersions.git;
+        };
       });
 
       defaultPackage = forAllSystems (system: self.packages.${system}.nix-serve);
 
-      checks = forAllSystems (system: {
+      checks = forAllSystems (system: let
+        pkgs = nixpkgs.legacyPackages.${system};
+      in {
         build = self.defaultPackage.${system};
-        # FIXME: add a proper test.
+      } // nixpkgs.lib.optionalAttrs (pkgs.stdenv.isLinux) {
+        nixos-test = pkgs.callPackage ./nixos-test.nix {
+          nix-serve = self.defaultPackage.${system};
+        };
```

Comment:
This is fine, but alternatively, you could make a scope, like the `nix` flake does, or as Pol has boiled down quite nicely [here](https://github.com/drupol/pkgs-by-name-for-flake-parts/blob/85c9b1712e269c8f756216d68cd769bb613d5ba6/flake-module.nix#L47).
(not to say you need flake-parts; just an example I came across today, and it's short)

Reply:
I used the suggestion from below.